src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Spark.Sql;
+using Xunit;
+using static Microsoft.Spark.Sql.Functions;
+
+namespace Microsoft.Spark.E2ETest.UdfTests
+{
+    [Collection("Spark E2E Tests")]
+    public class UdfComplexTypesTests
+    {
+        private readonly SparkSession _spark;
+        private readonly DataFrame _df;
+
+        public UdfComplexTypesTests(SparkFixture fixture)
+        {
+            _spark = fixture.Spark;
+            _df = _spark
+                .Read()
+                .Json(Path.Combine($"{TestEnvironment.ResourceDirectory}people_types.json"));
+        }
+
+        /// <summary>
+        /// UDF that takes in Array type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithArrayType()
+        {
+            // ArrayList works for this type.
+            Func<Column, Column> udfInt = Udf<int[], string>(array => string.Join(',', array));
+
+            Assert.Throws<Exception>(() => _df.Select(udfInt(_df["ages"])).Collect().ToArray());
+
+            Func<Column, Column> udfString = Udf<string[], string>(
+                array => string.Join(',', array));
+
+            Assert.Throws<Exception>(() => _df.Select(udfString(_df["cars"])).Collect().ToArray());
+        }
+
+        /// <summary>
+        /// UDF that returns Array type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnTypeAsArray()
+        {
+            Func<Column, Column> udf = Udf<string, string[]>(
+                str => new string[] { str, str + str });
+
+            Assert.Throws<NotImplementedException>(
+                () => _df.Select(udf(_df["name"])).Collect().ToArray());
+        }
+
+        /// <summary>
+        /// UDF that takes in Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithMapType()
+        {
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    dict => dict.Count.ToString());
+
+            Assert.Throws<Exception>(() => _df.Select(udf(_df["info"])).Collect().ToArray());

```

Comment:
Can you add a comment what exception message is thrown on the worker side? Let's add a snippet of it.

Reply:
I have added the comments for MayType and ArrayType. Please let me know if this makes sense.
​​​​​​​